### PR TITLE
Fragmentation counters, for IPv4 and IPv6

### PR DESCRIPTION
**Pull request by kbara**
This also includes a facility to regenerate test counter files.

### CODE REVIEW

File: src/apps/lwaftr/fragmentv4_hardened.lua [hunk 1]
```diff
@@ -267,20 +267,23 @@ function initialize_frag_table(max_fragmented_packets, max_pkt_frag)
...
 end
 
 function cache_fragment(frags_table, fragment)
    local key = get_key(fragment)
    local ptr = frags_table:lookup_ptr(key)
+   local ej = false
    if not ptr then
       local reassembly_buf = packet_to_reassembly_buffer(fragment)
-      frags_table:add_with_random_ejection(key, reassembly_buf, false)
+      _, ej = frags_table:add_with_random_ejection(key, reassembly_buf, false)
```
Comment (dpino) [inline src/apps/lwaftr/fragmentv4_hardened.lua:282]: Or can use `select` to pick one of the returned values, for instance:

```
ej = select(2, frags_table:add_with_random_ejection(key, reassembly_buf, false))
```

File: src/program/lwaftr/check/check.lua [hunk 2]
```diff
@@ -75,6 +79,18 @@ function validate_diff(actual, expected)
    end
 end
 
+local function regen_counters(counters, outfile)
+   local cnames = {}
+   for k in pairs(counters) do table.insert(cnames, k) end
```
Comment (dpino) [inline src/program/lwaftr/check/check.lua]: lwutil has a keys function.

``` lua
local cnames = lwutil.keys(counters)
```

**Comment by dpino** [comment]
LGTM. Maybe is worth to wait for a second review from @teknico

File: src/apps/lwaftr/fragmentv4_hardened.lua (hunk 1 above)
Comment (kbara) [reply src/apps/lwaftr/fragmentv4_hardened.lua:282]: I could, but that strikes me as a lot less readable. Call me an Erlanger if you must. :-)

**Comment by kbara** [comment]
Definitely. Counter documentation changes are blocked on him sorting out the merge conflicts that have been committed to that file (by him, according to `git blame`.)

**Comment by kbara** [comment]
I've moved the 'line oriented' input/output/drop counters into the fragmentation and reassembly apps. The benefit is that they now more closely reflect what goes over the line. There are a number of tradeoffs involved. (To be entirely accurately line-oriented, the counters would have to be in the driver itself - for instance, we never count vlan tags, because they're already stripped - but I'd rather keep the counters card-agnostic.)

Conceptually, this belongs in a separate app, rather than the fragmentation and reassembly apps, but I don't want to add 4 extra apps to the app network. I've placed them where they are because it's at the outer fringes of our app network (excluding the card, and the software vlan tag handlers that we use for testing.)

**Comment by teknico** [comment]
Oops, sorry, I wonder how that happened. Making a PR to remove the merge conflicts from `README.counters.md`.

File: src/apps/lwaftr/fragmentv4_hardened.lua (hunk 1 above)
Comment (teknico) [reply src/apps/lwaftr/fragmentv4_hardened.lua:282]: Python and Go use the same convention, it's not just Erlang. :-)

File: src/program/lwaftr/check/check.lua [hunk 3]
```diff
@@ -8,8 +8,11 @@ local setup = require("program.lwaftr.setup")
...
-local counter_names = lwaftr.counter_names
-local counters_dir = lwaftr.counters_dir
+local lwcounter = require("apps.lwaftr.lwcounter")
+local counter_names = lwcounter.counter_names
+local counters_dir = lwcounter.counters_dir
+local lwutil = require("apps.lwaftr.lwutil")
+local write_to_file = lwutil.write_to_file
 
 function show_usage(code)
    print(require("program.lwaftr.check.README_inc"))
```
Comment (teknico) [inline src/program/lwaftr/check/check.lua:18]: The README file needs a description of the new `-r` option, and of the existing `on-a-stick` option too.

**Comment by teknico** [comment]
Will LGTM once the README is updated.

I wonder how to redraw the counter diagrams taking into account the counters' new place in the code.

**Pull request merged**